### PR TITLE
Improved greedy worker scheduling

### PR DESCRIPTION
**Pull request by grananqvist**
Now assigns user to worker with the smallest accumulated user weight so far, like in a textbook greedy scheduling algorithm. Before, because of implementation issues, scheduling based on weight was done, but had to select the same number of users per process.

### CODE REVIEW

File: benchmarks/dataset/argument_parsing.py [hunk 1]
```diff
@@ -132,6 +132,20 @@ def add_dataset_arguments(
                             default=100,
                             help='Maximum number of images per user')
 
+        parser.add_argument(
+            '--scheduling_base_weight_multiplier',
+            type=float,
+            default=1.0,
+            help=('Figure 3b in pfl-research paper '
+                  'https://arxiv.org/abs/2404.06430 show adding a'
```
Comment (grananqvist) [inline benchmarks/dataset/argument_parsing.py:140]: paper not yet updated

File: benchmarks/dataset/flair/numpy.py [hunk 2]
```diff
@@ -11,12 +12,15 @@
...
 
 def make_federated_dataset(
         hdf5_path: str,
         partition: str,
         use_fine_grained_labels: bool,
         max_num_user_images: int,
+        scheduling_base_weight_multiplier: int = 1,
         numpy_to_tensor: Callable = lambda x: x) -> FederatedDataset:
     """
     Create federated dataset from the flair dataset, to use in simulations.
```
Comment (ac554) [inline benchmarks/dataset/flair/numpy.py:26]: This function now also controls the scheduling of users through a greedy algorithm. This is not alluded to in this docstring.

File: benchmarks/dataset/flair/numpy.py [hunk 3]
```diff
@@ -37,8 +41,21 @@ def make_federated_dataset(
...
     """
     num_classes = len(get_label_mapping(hdf5_path, use_fine_grained_labels))
-    user_num_images = get_user_num_images(hdf5_path, partition)
-    user_ids = sorted(user_num_images.keys())
+    user_id_to_weight = {
+        k: min(v, max_num_user_images)
+        for k, v in get_user_num_images(hdf5_path, partition).items()
+    }
+    median_datapoints = np.median(list(user_id_to_weight.values()))
+    base_value = scheduling_base_weight_multiplier * median_datapoints
```
Comment (ac554) [inline benchmarks/dataset/flair/numpy.py:49]: This feels like new functionality which should be abstracted out of this function `make_federated_dataset`.

File: tests/data/test_federated_dataset.py [hunk 4]
```diff
@@ -266,7 +266,7 @@ def test_3_workers(self, mock_get_ops, federated_dataset_mixture,
 @pytest.fixture
 def user_id_to_weight(request):
     if hasattr(request, 'param') and request.param:
-        return {i: i for i in range(100)}
+        return {i: i + 1 for i in range(100)}
```
Comment (ac554) [inline tests/data/test_federated_dataset.py:269]: I presume the "1" is for base weight? Perhaps a comment, or creating a variable name, `base_weight=1`, would help with legibility.

Comment (grananqvist) [reply tests/data/test_federated_dataset.py:269]: this is just to avoid a user with weight 0

File: benchmarks/dataset/flair/numpy.py [hunk 5]
```diff
@@ -11,12 +12,15 @@
...
 
+logger = logging.getLogger(name=__name__)
+
 
 def make_federated_dataset(
         hdf5_path: str,
         partition: str,
         use_fine_grained_labels: bool,
         max_num_user_images: int,
+        scheduling_base_weight_multiplier: int = 1,
```
Comment (congzheng-song) [inline benchmarks/dataset/flair/numpy.py:23]: It is a float right? from the argparse above

**Review by congzheng-song — COMMENTED** [review]
Some minor comments

File: benchmarks/dataset/flair/numpy.py (hunk 3 above)
Comment (congzheng-song) [reply benchmarks/dataset/flair/numpy.py:49]: Right, if you think this (scaled median) can generally benefits most datasets. But if we only have evidence on FLAIR, I think it is ok to leave it here for now.

**Review by ac554 — APPROVED** [review]

**Pull request merged**